### PR TITLE
CSV output, I/O, IDD parser, and file conversion bugfixes

### LINKED ISSUE #8208: --convert-only option does not exit properly - continues with setup and sizing simulation
Issue overview
--------------
Using the `--convert-only` command-line option should convert between idf and epJSON and exit without doing any simulations. Current develop shows that it is continuing with initialization and sizing simulations. 9.3.0 was also doing this silently (output files contained sizing results).

The problem is [here in EnergyPlusPgm.cc](https://github.com/NREL/EnergyPlus/blob/5bf36f34bad6254a899b749d0c76a9d072c25124/src/EnergyPlus/api/EnergyPlusPgm.cc#L428-L438).
```
    int status = initializeAsLibrary(state);
    if (status) return status;
    try {
        EnergyPlus::SimulationManager::ManageSimulation(state);
    } catch (const EnergyPlus::FatalError &e) {
        return EnergyPlus::AbortEnergyPlus(state);
    } catch (const std::exception &e) {
        EnergyPlus::ShowSevereError(e.what());
        return EnergyPlus::AbortEnergyPlus(state);
    }
    return wrapUpEnergyPlus(state);
```

`initializeAsLibrary` returns `EXIT_SUCCESS` which is zero.

Per @mbadams5 it isn’t an error to end early….  so we might need another status code.

### Details
Some additional details for this issue (if relevant):
 - Platform Windows 10
 - Version of EnergyPlus 9.4.0-95f55a4710

Here's the console output:
```
energyplus.exe --convert-only "5ZoneAirCooled - Copy.idf"
EnergyPlus Starting
EnergyPlus, Version 9.4.0-95f55a4710, YMD=2020.08.12 13:21
Converted input file format. Exiting.
EnergyPlus Run Time=00hr 00min  7.51sec
EnergyPlus Completed Successfully.
Initializing Response Factors
Calculating CTFs for "FLOOR-SLAB-1"
Calculating CTFs for "INT-WALL-1"
Calculating CTFs for "ROOF-1"
Calculating CTFs for "WALL-1"
Initializing Window Optical Properties
Initializing Solar Calculations
Allocate Solar Module Arrays
Initializing Zone and Enclosure Report Variables
Initializing Surface (Shading) Report Variables
Computing Interior Solar Absorption Factors
Determining Shadowing Combinations
Computing Window Shade Absorption Factors
Proceeding with Initializing Solar Calculations
Initializing Surfaces
Initializing Outdoor environment for Surfaces
Setting up Surface Reporting Variables
Initializing Temperature and Flux Histories
Initializing Window Shading
Computing Interior Absorption Factors
Computing Interior Diffuse Solar Absorption Factors
Initializing Solar Heat Gains
Initializing Internal Heat Gains
Initializing Interior Solar Distribution
Initializing Interior Convection Coefficients
Gathering Information for Predefined Reporting
Completed Initializing Surface Heat Balance
Calculate Outside Surface Heat Balance
Calculate Inside Surface Heat Balance
Calculate Air Heat Balance
Initializing HVAC
Warming up
Warming up
Warming up
Warming up
Warming up
Warming up
Performing Zone Sizing Simulation for Load Component Report
...for Sizing Period: #1 CHICAGO_IL_USA ANNUAL COOLING 1% DESIGN CONDITIONS DB/MCWB
Warming up
Warming up
Warming up
Warming up
Warming up
Warming up
Performing Zone Sizing Simulation for Load Component Report
...for Sizing Period: #2 CHICAGO_IL_USA ANNUAL HEATING 99% DESIGN CONDITIONS DB
Re-zeroing zone sizing arrays
Warming up
Warming up
Warming up
Warming up
Warming up
Warming up
Performing Zone Sizing Simulation
...for Sizing Period: #1 CHICAGO_IL_USA ANNUAL COOLING 1% DESIGN CONDITIONS DB/MCWB
Warming up
Warming up
Warming up
Warming up
Warming up
Warming up
Performing Zone Sizing Simulation
...for Sizing Period: #2 CHICAGO_IL_USA ANNUAL HEATING 99% DESIGN CONDITIONS DB
Calculating System sizing
...for Sizing Period: #1 CHICAGO_IL_USA ANNUAL COOLING 1% DESIGN CONDITIONS DB/MCWB
Calculating System sizing
...for Sizing Period: #2 CHICAGO_IL_USA ANNUAL HEATING 99% DESIGN CONDITIONS DB
Adjusting Air System Sizing
Adjusting Standard 62.1 Ventilation Sizing
Initializing Simulation
Reporting Surfaces
Beginning Primary Simulation
Writing final SQL reports
**FATAL:Error condition occurred.  Previous Severe Errors cause termination.
```

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### LINKED ISSUE #8223: Infinite loop in IDD->Schema parser
Issue overview
--------------
Helping diagnose a problem with an infinitely long running schema generation step, I found that the culprit was a field with `\type object-lis` instead of `\type object-list`.  I debugged and found the parser continually trying to parse a token that started with `o` over and over infinitely.  I verified by taking our own IDD and just replacing one of the type declarations, and it demonstrates the problem.

### LINKED ISSUE #8252: csv output problems
Issue overview
--------------
There are some problems with eplusout.csv and eplusmtr.csv when generated directly from EnergyPlus using OutputControl:Files "Output CSV" = Yes:

* On windows - The header line has an extra CR character (OD OD OA on windows). Don't know if there's an extra character on other platforms.
* Monthly output for eplusmtr.csv is broken. The values are misaligned and some values are missing. (Hourly meter csv is OK. and Monthly output variable csv is OK). Need to test all reporting frequencies for both files and test mixed reporting frequencies.

### Details
Some additional details for this issue (if relevant):
 - Platform Windows 10
 - Version of EnergyPlus 9.4.0-70377e364c
 - Some of these problems were noted by @Myoldmopar in https://github.com/NREL/EnergyPlus/pull/7904#issuecomment-678405079

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect files: 
[8252-OutputFiles.zip](https://github.com/NREL/EnergyPlus/files/5170470/8252-OutputFiles.zip)

 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### LINKED ISSUE #8279: WeatherManager Hangs on RunPeriods that Run Past the End of Year
Issue overview
--------------
Need to address the second part of #8238. ```utility_bill02.rb.idf``` has a RunPeriod that runs April to April. WeatherManager drops into an infinite loop after reading the last line of the weather file. This comment can provide some additional context.
https://github.com/NREL/EnergyPlus/issues/8238#issuecomment-692320703

### Details
Some additional details for this issue (if relevant):
 - Platform (Mac 10.14.6)
 - Version of EnergyPlus (ca23a81039ec91b3506b2c22ae4f90b794e662a8)
 - Unmethours link or helpdesk ticket number - None

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by mbadams5**
Pull request overview
---------------------
 - Fixes #8223
 - Fixes #8208
 - Fixes #8252
 - Fixes #8279
 - Fixes [this comment](https://github.com/NREL/EnergyPlus/pull/7904#discussion_r485231814)

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by Myoldmopar** [comment]
This is a great collection of fixes, thanks @mbadams5.  I'll check them out tomorrow when I can, but if @mitchute or @mjwitte could attempt to verify some of them too that would be ok.

File: scripts/dev/generate_epJSON_schema/idd_parser.py [hunk 1]
```diff
@@ -370,13 +370,14 @@ def parse_field(data, token):
...
-
             elif match_string(data, REAL_STR) or match_string(data, INTEGER_STR):
                 if 'type' not in root or 'type' != 'number':
                     root['type'] = 'number'
-
             elif match_string(data, NODE_STR):
                 root['type'] = 'string'
+            else:
+                bad_type = parse_line(data)
+                raise RuntimeError("Invalid \\type: \"%s\"" % bad_type)
```
Comment (Myoldmopar) [inline scripts/dev/generate_epJSON_schema/idd_parser.py:380]: So this should be catching the infinite loop on a bad `type` declaration in the IDD parser.  :+1:

File: src/EnergyPlus/InputProcessing/InputProcessor.cc [hunk 2]
```diff
@@ -283,7 +283,7 @@ void InputProcessor::processInput()
             std::string input_file;
             std::string line;
             while (std::getline(input_stream, line)) {
-                input_file.append(line + DataStringGlobals::NL);
+                input_file.append(line + '\n');
```
Comment (Myoldmopar) [inline src/EnergyPlus/InputProcessing/InputProcessor.cc:286]: Cleaning `NL` back out :+1:

File: src/EnergyPlus/api/EnergyPlusPgm.cc [hunk 3]
```diff
@@ -396,7 +396,7 @@ int RunEnergyPlus(EnergyPlus::EnergyPlusData &state, std::string const & filepat
     // as possible and contain all "simulation" code in other modules and files.
 
     int status = initializeEnergyPlus(state, filepath);
-    if (status) return status;
+    if (status || EnergyPlus::DataGlobals::outputEpJSONConversionOnly) return status;
```
Comment (Myoldmopar) [inline src/EnergyPlus/api/EnergyPlusPgm.cc:399]: Allow the simulation to exit gracefully when just converting :+1:

File: src/EnergyPlus/OutputProcessor.cc [hunk 4]
```diff
@@ -3338,9 +3338,6 @@ namespace OutputProcessor {
         if (!ResultsFramework::resultsFramework->DYMeters.rVariablesScanned()) {
             ResultsFramework::resultsFramework->initializeMeters(EnergyMeters, ReportingFrequency::Daily);
         }
-        if (ResultsFramework::resultsFramework->DYMeters.rDataFrameEnabled()) {
-            ResultsFramework::resultsFramework->DYMeters.newRow(Month, DayOfMonth, HourOfDay, 0);
-        }
```
Comment (Myoldmopar) [inline src/EnergyPlus/OutputProcessor.cc:3343]: I guess this fixes the meter alignment issues.

File: src/EnergyPlus/IOFiles.hh [hunk 5]
```diff
@@ -100,7 +100,12 @@ public:
     void open(bool = false, bool = true);
     std::fstream::pos_type position() const noexcept;
 
-    void rewind() noexcept { if (is) { is->seekg(0); } }
+    void rewind() noexcept {
+        if (is) {
+            is->clear(); // clear eofbit and potentially failbit
+            is->seekg(0, std::ios::beg);
+        }
+    }
```
Comment (Myoldmopar) [inline src/EnergyPlus/IOFiles.hh:108]: And I suppose this is the fix for the weather manager hang.

**Review by Myoldmopar — APPROVED** [review]
These are some pretty great fixes.  I'm building locally to see if I can deduce the failing unit test issue.

File: src/EnergyPlus/OutputProcessor.cc (hunk 4 above)
Comment (mbadams5) [reply src/EnergyPlus/OutputProcessor.cc:3343]: The fixes like this and the change from Hourly to Yearly for default, yes. I didn't do a thorough testing of all combinations, but monthly works properly now. I tested with 5ZoneAirCooled.idf. I am fairly confident these fix all the issues, but if someone else could double check that would be good.

File: src/EnergyPlus/IOFiles.hh (hunk 5 above)
Comment (mbadams5) [reply src/EnergyPlus/IOFiles.hh:108]: Yes, it is. Each commit is atomic to the bug fix it to attempting to solve.

File: src/EnergyPlus/api/EnergyPlusPgm.cc (hunk 3 above)
Comment (Myoldmopar) [reply src/EnergyPlus/api/EnergyPlusPgm.cc:399]: I tested this and it works well, thanks!

**Comment by mitchute** [comment]
Confirmed that this corrects #8279

**Comment by Myoldmopar** [comment]
This is looking really great.  All 3 platforms pass successfully.  I think I'll go ahead and merge it in now rather than waste another integration debug build on it.  Thanks for confirming that bug fix @mitchute .  Thanks for all these fixes @mbadams5 .

**Pull request merged**